<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Deriving reference profiles from scRNA-seq data</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Deriving reference profiles from scRNA-seq
data</h1>



<style>
p.caption {
font-size: 1.5em;
}
</style>
<p>Here we’ll show how to derive reference profiles from a scRNA-seq
dataset. These profiles can then be used to run Insitutype in supervised
or semi-supervised mode.</p>
<p>First, we load the package and access the example data.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(InSituType)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;mini_nsclc&quot;</span>)</span></code></pre></div>
<p>You’ll have a scRNA-seq dataset in whatever format you prefer. For
our tools, you’ll just need to extract the raw, linear-scale expression
matrix and a vector of cell type assignments.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># pretend the demo cosmx data is your scRNA-seq counts matrix</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>counts <span class="ot">&lt;-</span> mini_nsclc<span class="sc">$</span>counts  </span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">dim</span>(counts)  <span class="co"># cells in rows, genes in columns</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co">#&gt; [1] 2198  960</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>counts[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]  <span class="co"># see linear scale</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co">#&gt;          AATK ABL1 ABL2 ACE ACE2</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co">#&gt; c_3_18_2    0    0    0   0    0</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co">#&gt; c_3_18_3    0    0    0   0    0</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co">#&gt; c_3_18_4    0    0    0   0    0</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#&gt; c_3_18_5    0    0    0   0    0</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#&gt; c_3_18_6    0    0    0   0    0</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="co"># make up a vector of cell type assignments:</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>clust <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">c</span>(<span class="st">&quot;a&quot;</span>,<span class="st">&quot;b&quot;</span>,<span class="st">&quot;c&quot;</span>), <span class="fu">nrow</span>(counts), <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="fu">table</span>(clust)</span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="co">#&gt; clust</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co">#&gt;   a   b   c </span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="co">#&gt; 748 710 740</span></span></code></pre></div>
<p>Now run the getRNAprofiles function:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>profiles <span class="ot">&lt;-</span> <span class="fu">getRNAprofiles</span>(<span class="at">x =</span> counts, </span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>                           <span class="at">neg =</span> <span class="dv">0</span>,   <span class="co"># assume no background in scRNA-seq data. </span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>                           <span class="at">clust =</span> clust) </span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="fu">head</span>(profiles)</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co">#&gt;                c           a          b</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co">#&gt; AATK  0.01081081 0.009358289 0.01126761</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co">#&gt; ABL1  0.02702703 0.029411765 0.03380282</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co">#&gt; ABL2  0.05675676 0.070855615 0.06197183</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co">#&gt; ACE   0.02972973 0.049465241 0.02676056</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="co">#&gt; ACE2  0.02432432 0.026737968 0.03239437</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="co">#&gt; ACKR1 0.10135135 0.081550802 0.07887324</span></span></code></pre></div>
<p>And of course, if you were deriving a profile matrix from a previous
CosMx dataset or any other technology with background, you’d want to
enter the per-cell background level in the “neg” argument.</p>
<pre><code>


### Session Info


```r
sessionInfo()
#&gt; R version 4.2.2 (2022-10-31 ucrt)
#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)
#&gt; Running under: Windows 10 x64 (build 19045)
#&gt; 
#&gt; Matrix products: default
#&gt; 
#&gt; locale:
#&gt; [1] LC_COLLATE=English_United States.utf8 
#&gt; [2] LC_CTYPE=English_United States.utf8   
#&gt; [3] LC_MONETARY=English_United States.utf8
#&gt; [4] LC_NUMERIC=C                          
#&gt; [5] LC_TIME=English_United States.utf8    
#&gt; 
#&gt; attached base packages:
#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     
#&gt; 
#&gt; other attached packages:
#&gt; [1] InSituType_1.0.0
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;  [1] mclust_6.0.1                Rcpp_1.0.9                 
#&gt;  [3] lattice_0.20-45             deldir_1.0-6               
#&gt;  [5] png_0.1-8                   digest_0.6.31              
#&gt;  [7] SingleCellExperiment_1.20.1 utf8_1.2.2                 
#&gt;  [9] RSpectra_0.16-1             lsa_0.73.3                 
#&gt; [11] R6_2.5.1                    GenomeInfoDb_1.34.9        
#&gt; [13] stats4_4.2.2                evaluate_0.23              
#&gt; [15] ggplot2_3.4.4               pillar_1.9.0               
#&gt; [17] zlibbioc_1.44.0             rlang_1.1.3                
#&gt; [19] rstudioapi_0.15.0           data.table_1.14.6          
#&gt; [21] irlba_2.3.5.1               jquerylib_0.1.4            
#&gt; [23] S4Vectors_0.36.1            Matrix_1.6-5               
#&gt; [25] reticulate_1.34.0           rmarkdown_2.25             
#&gt; [27] uwot_0.1.14                 RCurl_1.98-1.9             
#&gt; [29] polyclip_1.10-4             munsell_0.5.0              
#&gt; [31] umap_0.2.10.0               DelayedArray_0.24.0        
#&gt; [33] spatstat.data_3.0-4         compiler_4.2.2             
#&gt; [35] xfun_0.41                   askpass_1.2.0              
#&gt; [37] pkgconfig_2.0.3             BiocGenerics_0.44.0        
#&gt; [39] htmltools_0.5.7             openssl_2.1.1              
#&gt; [41] tidyselect_1.2.0            SummarizedExperiment_1.28.0
#&gt; [43] tibble_3.1.8                GenomeInfoDbData_1.2.9     
#&gt; [45] IRanges_2.32.0              matrixStats_0.63.0         
#&gt; [47] fansi_1.0.3                 dplyr_1.0.10               
#&gt; [49] bitops_1.0-7                SnowballC_0.7.1            
#&gt; [51] grid_4.2.2                  jsonlite_1.8.8             
#&gt; [53] gtable_0.3.4                lifecycle_1.0.4            
#&gt; [55] DBI_1.2.1                   magrittr_2.0.3             
#&gt; [57] scales_1.3.0                cli_3.5.0                  
#&gt; [59] cachem_1.0.6                XVector_0.38.0             
#&gt; [61] bslib_0.6.1                 generics_0.1.3             
#&gt; [63] vctrs_0.5.2                 spatstat.utils_3.0-4       
#&gt; [65] tools_4.2.2                 Biobase_2.58.0             
#&gt; [67] glue_1.6.2                  MatrixGenerics_1.10.0      
#&gt; [69] fastmap_1.1.1               yaml_2.3.6                 
#&gt; [71] colorspace_2.0-3            GenomicRanges_1.50.2       
#&gt; [73] spatstat.geom_3.1-0         knitr_1.45                 
#&gt; [75] sass_0.4.8</code></pre>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
