---
title: "Clustering a small CosMx RNA dataset with Insitutype"
output: 
  rmarkdown::html_vignette: 
    toc: true
fig_width: 10 
fig_height: 6 
vignette: >
  %\VignetteIndexEntry{Clustering a small CosMx RNA dataset with Insitutype}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
<style>
p.caption {
  font-size: 1.5em;
}
</style>
```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
### Installation
```{r installation, eval=FALSE}

# Install remotes package if necessary
if (!require("remotes")) {
  install.packages("remotes")
}

remotes::install_github("Nanostring-Biostats/InSituType")

```   
### Overview
This vignette demonstrates the use of the Insitutype algorithm to cluster CosMx RNA data. We'll analyze a small subset of the non-small cell lung cancer (NSCLC) 1000-plex public data release.

For a deep-dive in the nuances of running Insitutype and QC-ing its results, see the FAQS.md file. 

### Data preparation
First, we load the package and access the example data. The NSCLC RNA data R object is a list including RNA expression (`counts`) and negative control probe expression (`neg`) in `mini_nsclc` R data file. `ioprofiles` is a matrix holding reference cell type expression profiles. 

```{r setup}
library(InSituType)
data("ioprofiles")
data("iocolors")
data("mini_nsclc")

str(mini_nsclc)
head(ioprofiles)
set.seed(0)
```

#### Necessary inputs
First, let us look at the data we need.  
1. A matrix of RNA data, with cells in rows and genes in columns. Because Insitutype is based in negative binomial likelihoods, RNA data must be integer-valued. 
```{r strcounts}
counts <- mini_nsclc$counts
str(counts)
counts[25:30, 9:14]

```

2. A vector giving each cell's mean negative control value. The example data includes `mini_nsclc$neg`, which is a data frame. Hence, we need to estimate average values of the negative probes.  
```{r strnegmean}
negmean <- Matrix::rowMeans(mini_nsclc$neg)
head(negmean)
```

3. Optionally, a vector of each cell's expected background counts (per gene). If not provided, Insitutype will estimate this value from the negative controls.

4. For supervised or semi-supervised mode, a reference profile matrix giving the expected expression profile of each cell type, with genes in rows and cell types in columns. 
The reference matrix must be linear-scale, not log-scale. Insitutype will automatically align the gene in this matrix to your expression data. 

For supervised cell typing, reference matrices should contain every cell type present in the tissue. If you suspect the tissue has cell types outside your reference matrix, use Insitutype's semi-supervised clustering capability to detect new clusters. 


#### Incorporating additional data types
Single-cell spatial transcriptomics datasets offer data types beyond simple gene expression, from single-cell images to spatial contexts. Insitutype digests this information in the form of "cohorts" of cells with similar data from the alternative data types. Then, when assigning cell types, Insitutype will considers the frequency of each cell type in each cohort. 

For example, if you have a cohort of mainly CD45- cells, Insitutype will see that immune cells are very uncommon within this cohort, and it will give cells in this cohort lower probabilities of falling in immune cell clusters. (Insitutype doe not know about CD45 staining; it just considers the 
frequency of each cluster in each cohort.)

Cohort definition is left to the user, and many approaches are reasonable. 
For example, you might:

* Define cohort by gating cells on their immunofluorescence stains, e.g. PanCK+/- or CD45+/-
* Cluster cells on their continuous immunofluorescence values
* Use an autoencoder to extract features from cells images, then cluster on those features. 
* Use the results of spatial clustering / "niches" to define cohorts
* Cluster on information from multiple data types together, such as the average expression profile of cell neighbors AND immunofluorescence. 

Cohorting is a low-risk undertaking, so it is recommended whenever possible. A poorly-defined cohort will simply be uninformative; it will not send the clustering algorithm astray. The more cohorts you can define, and the higher-quality those cohorts, the more information Insitutype will be able to gleam from the alternative data types. 

One warning: **do not define cohorts based on variables you want to study.**
For example, if you wish to study how cell populations shift across spatial context, then spatial context must not be allowed to inform cell typing. To do so would introduce circularity. 

In the below example, we will define cohorts using the "fastCohorting" function. This function is provided as a convenient way to derive cohorts from diverse data types; it makes no claim to being optimal. 

```{r cohorting, echo = TRUE}
# simulate immunofluorescence data: 
immunofluordata <- matrix(rpois(n = nrow(counts) * 4, lambda = 100), nrow(counts))
# perform automatic cohorting:
cohort <- fastCohorting(immunofluordata, gaussian_transform = TRUE) 
# ("Gaussian_transform = TRUE" maps variables to Gaussians in order to 
#  place dramatically different variables on the same scale.)
table(cohort)
```

### Unsupervised clustering:

The below command performs unsupervised clustering. We specify a number of clusters with the `n_clusts` argument, and we set `reference_profiles` as NULL.  

```{r unsup, echo=TRUE}
unsup <- insitutype(
  x = counts,
  neg = negmean,
  assay_type = "RNA", 
  reference_profiles = NULL,
  reference_sds = NULL,
  cohort = cohort,
  n_clusts = 10
) 
```

### Supervised cell typing:

The below command performs supervised cell typing The user needs to provide `reference_profiles` and set `n_clusts=0`. 
For supervised cell typing, there is usually no need to update the reference profile, though if your initial results are
unsatisfying then setting `update_reference_profiles = FALSE` can help. 


```{r sup, echo=TRUE}
sup <- insitutype(
  x = counts,
  neg = negmean,
  assay_type = "RNA", 
  n_clusts=0, 
  anchors = NULL, 
  reference_profiles = ioprofiles, 
  cohort = cohort,
  update_reference_profiles = FALSE
) 
```

### Semi-supervised clustering:

The below command performs semi-supervised clustering.
For semi-supervised clustering, it is usually necessary to update your reference profiles to better fit CosMx data.
Insitutype uses "anchor" cells with high-confidence cell types to perform this calibration. 
You can specify anchors yourself, or, as below, you can let the algorithm choose them automatically.
For semi-supervised clustering, we specify a non-zero `n_clusts` value and supply a reference profiles matrix. 

```{r semisup, echo=TRUE}
semi_sup <- insitutype(
  x = counts,
  neg = negmean,
  assay_type = "RNA", 
  n_clusts = 3, 
  anchors = NULL, 
  reference_profiles = ioprofiles, #[names(sig_mat)[-c(1:2)],],
  cohort = cohort,
  update_reference_profiles = TRUE)
```

The function returns 4 useful outputs: 

* A vector of cells cluster assignments, 
* A vector of the posterior probability for each cell typing call,
* A matrix of cells log-likelihoods under each cluster, 
* A matrix of the mean background-subtracted expression profile of each cluster. 
* If present, a vector specifying the anchor cells


```{r str}
str(semi_sup)
```

### Visualizing clustering results

The below plots provide a good first look at your results:

```{r unsup_results, fig.width = 8, fig.height = 5}
## what do the cluster profiles look like?
heatmap(sweep(unsup$profiles, 1, pmax(apply(unsup$profiles, 1, max), .2), "/"), scale = "none",
        main = "Cluster mean expression profiles")
## where are the clusters in xy and umap space?
# define cluster colors:
cols <- c('#8DD3C7','#BEBADA','#FB8072','#80B1D3','#FDB462','#B3DE69','#FCCDE5',
    '#D9D9D9','#BC80BD','#CCEBC5','#FFED6F','#E41A1C','#377EB8','#4DAF4A',
    '#984EA3','#FF7F00','#FFFF33','#A65628','#F781BF','#999999')
cols <- cols[seq_along(unique(unsup$clust))]
names(cols) <- unique(unsup$clust)
# plot in umap and xy:
par(mfrow = c(1, 2))
par(mar = c(0, 0, 3, 0))
plot(mini_nsclc$x, mini_nsclc$y, pch = 16, cex = .75, asp = 1, cex.main = 0.75,
          main = "cells in physical space",
     col = cols[unsup$clust], xlab = "", ylab = "", xaxt = "n", yaxt = "n")
plot(mini_nsclc$umap, pch = 16, cex = .75, asp = 1, cex.main = 0.75,
     main = "cells in UMAP space",     
     col = cols[unsup$clust], xlab = "", ylab = "", xaxt = "n", yaxt = "n")
legend("bottomleft", pch = 16, col = cols, legend = names(cols), cex = 0.7)


```

Insitutype calculates the probability of each cell belonging to each cluster. This information lets us understand how clusters relate to each other. 
The flightpath_plot function arrays cells according to their cluster probabilities:

```{r flightpath, fig.width = 5, fig.height = 5}
# make the flightpath plot
fp <- flightpath_plot(flightpath_result = NULL, insitutype_result = unsup, col = cols[unsup$clust])
class(fp)
print(fp)
```

In the above plot, a cell atop a cluster centroid has near-certain cell type, while cells halfway between two centroids have 50% posterior probability of belonging to each of those two clusters. The number atop each centroid gives the average posterior probability of cells in the cluster. Expect to see this value above 0.9 for clustering major cell types, and above 0.75 when clustering apart closely-related cell types. Clusters with low average posterior probabilities are unreliable. You can either interpret them with caution or you can dissolve them and reassign their cells to other clusters. (The next section covers this operation.)

You can also extract the flightpath layout for plotting on your own:

```{r fp_layout}
# compute the flightpath layout:
fp_layout <- flightpath_layout(logliks = unsup$logliks, profiles = unsup$profiles)
str(fp_layout)

# plot the flightpath by hand:
par(mar = c(0,0,0,0))
plot(fp_layout$cellpos, pch = 16, cex = 0.5, col = cols[unsup$clust], xaxt = "n", yaxt = "n", xlab = "", ylab = "")
text(fp_layout$clustpos, rownames(fp_layout$clustpos), cex = 0.5)
```


#### Updating clustering results 

After close review, you might want to update your clustering results. 

The function refineClusters can modify your results in the following ways:

* Merge closely-related or frequently-confused clusters 
* Delete clusters (Sometimes a "catch-all" cluster arises that is frequently confused with multiple other clusters. It is best to delete these clusters and let their cells get reassigned to meaningful clusters.)
* Sub-cluster large or diverse clusters

Here we use the refineClusters function to perform all these operations. We also use its "merges" argument to rename cluster `c`. 

```{r refineClusters, fig.width = 5, fig.height = 5}
# refine the clusters:
newclusts <- refineClusters(logliks = unsup$logliks, 
                            merges = c("g" = "newcluster", "e" = "newcluster", "c" = "c_renamed"), 
                            to_delete = "a",
                            subcluster = c("f" = 2),
                            counts = counts,
                            neg = negmean) 
str(newclusts)

# plot the updated results:
cols <- c('#8DD3C7','#BEBADA','#FB8072','#80B1D3','#FDB462','#B3DE69','#FCCDE5',
    '#D9D9D9','#BC80BD','#CCEBC5','#FFED6F','#E41A1C','#377EB8','#4DAF4A',
    '#984EA3','#FF7F00','#FFFF33','#A65628','#F781BF','#999999')
cols <- cols[seq_along(unique(newclusts$clust))]
names(cols) <- unique(newclusts$clust)
fp <- flightpath_plot(flightpath_result = NULL, insitutype_result = newclusts, col = cols[newclusts$clust])
print(fp)
```


### Coloring cell types

The colorCellTypes function attempts to choose good colors for your cell types. Some of its options use the frequency of cell types in this decision. 
Here are 3 versions:

```{r colorCellTypes, fig.width = 8, fig.height = 5}


par(mfrow = c(1, 3))
par(mar = c(0, 0, 3, 0))

plot(mini_nsclc$x, mini_nsclc$y, pch = 16, cex = .75, asp = 1, cex.main = 0.75,
          main = "brewers palette",
     col = colorCellTypes(freqs = table(unsup$clust), palette = "brewers")[unsup$clust], 
     xlab = "", ylab = "", xaxt = "n", yaxt = "n")

plot(mini_nsclc$x, mini_nsclc$y, pch = 16, cex = .75, asp = 1, cex.main = 0.75,
          main = "tableau20 palette",
     col = colorCellTypes(freqs = table(unsup$clust), palette = "tableau20")[unsup$clust], 
     xlab = "", ylab = "", xaxt = "n", yaxt = "n")

plot(mini_nsclc$x, mini_nsclc$y, pch = 16, cex = .75, asp = 1, cex.main = 0.75,
          main = "earthplus palette",
     col = colorCellTypes(freqs = table(unsup$clust), palette = "earthplus")[unsup$clust], 
     xlab = "", ylab = "", xaxt = "n", yaxt = "n")

```

### Session Info

```{r sessioninfo}
sessionInfo()
```
